### PR TITLE
WIP: Fix accessibility issues for dropdown and tab components

### PR DESCRIPTION
**Pull request by hakjeri**
Not finished yet, but mainly cleanup and minor things left to do in my opinion. Good enough for initial review

### CODE REVIEW

File: src/components/Dropdown/Dropdown.tsx [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+  }
+
+  const deselectItem = (selectedItem: Item) =>
+    setItems((previousState) =>
+      previousState.filter((item) => item.id !== selectedItem.id)
+    )
+
+  const resultsToShow = Math.min(matchingItems.length, maxAmountOfItems)
+
+  const handleKeyDown: KeyboardEventHandler<HTMLInputElement> = (e) => {
```
Comment (tasola) [inline src/components/Dropdown/Dropdown.tsx:109]: Nice

File: src/components/Dropdown/Dropdown.tsx [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+            searchIsEnabled={enableSearch}
+            role='combobox'
+            aria-autocomplete='list'
+            aria-haspopup='listbox'
+            aria-expanded={doShowResultsList}
+            aria-activedescendant={
+              suggestionIndex === -1
+                ? undefined
+                : `${id}-option-${suggestionIndex}`
+            }
```
Comment (tasola) [inline src/components/Dropdown/Dropdown.tsx:175]: Interesting aria fields.

File: src/components/Dropdown/Dropdown.tsx [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+            }`}
+            placeholder={placeholder}
+            onChange={handleOnChange}
+            onKeyDown={handleKeyDown}
+            value={enableSearch ? undefined : placeholder}
+            onClick={() => {
+              setDoShowResultsList(!doShowResultsList)
+            }}
+            aria-controls={`${id} ${controls}`}
+            aria-describedby={`${id}-help`}
```
Comment (tasola) [inline src/components/Dropdown/Dropdown.tsx:165]: Why -help?

File: src/components/Dropdown/Dropdown.tsx [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+      const isGreaterThanResultsListLength = suggestionIndex < resultsToShow - 1
+      setSuggestionIndex(
+        isGreaterThanResultsListLength ? suggestionIndex + 1 : 0
+      )
+    }
+    if (e.key === 'ArrowUp' || e.key === 'Up') {
+      e.preventDefault()
+      setDoShowResultsList(true)
+      if (suggestionIndex === -1) {
+        setSuggestionIndex(0)
```
Comment (tasola) [inline src/components/Dropdown/Dropdown.tsx:138]: Fair enough. Is this some kind of recommendation to go "down" to the first item in the list by pressing 'up'?

File: src/pages/teaching-material/index.tsx [hunk 5]
```diff
@@ -65,13 +124,96 @@ export default function Discover() {
...
+        : ''
+    const isFiltered = keywordPart || authorPart
+    const fullSummary = `${dataResults} ${
+      isFiltered ? 'Filtered by' : ''
+    } ${keywordPart} ${authorPart ? 'and' : ''} ${authorPart}`
+    return fullSummary
+  }, [
+    // courseResults,
+    // lectureResults,
+    // blockResults,
```
Comment (tasola) [inline src/pages/teaching-material/index.tsx:147]: the above are not needed right?

File: src/pages/teaching-material/index.tsx [hunk 6]
```diff
@@ -65,13 +124,96 @@ export default function Discover() {
...
+
+    setResults({
+      courses: courseFilterResult,
+      lectures: lectureFilterResult,
+      blocks: blockFilterResult,
+    })
+
+    // setLectureResults(lectureFilterResult)
+    // setCourseResults(courseFilterResult)
+    // setBlockResults(blockFilterResult)
```
Comment (tasola) [inline src/pages/teaching-material/index.tsx:196]: Remove

File: src/pages/teaching-material/index.tsx [hunk 7]
```diff
@@ -39,9 +52,55 @@ const H2 = styled.h2`
...
+
+const defaultFilterResult: ResponseArrayData<any> = {
+  data: [],
+  meta: {
+    pagination: defaultPagination,
+  },
+}
+
 export default function Discover() {
+  // return null
```
Comment (tasola) [inline src/pages/teaching-material/index.tsx:73]: Remove

File: src/pages/teaching-material/index.tsx [hunk 8]
```diff
@@ -39,9 +52,55 @@ const H2 = styled.h2`
...
+    courses: defaultFilterResult,
+    lectures: defaultFilterResult,
+    blocks: defaultFilterResult,
+  })
+  const [courseResults, setCourseResults] =
+    useState<ResponseArrayData<CourseThreeLevelsDeep>>(defaultFilterResult)
+  const [lectureResults, setLectureResults] =
+    useState<ResponseArrayData<LectureTwoLevelsDeep>>(defaultFilterResult)
+  const [blockResults, setBlockResults] =
+    useState<ResponseArrayData<BlockOneLevelDeep>>(defaultFilterResult)
```
Comment (tasola) [inline src/pages/teaching-material/index.tsx:101]: Aren't these states replaced by `results`?

File: src/pages/teaching-material/index.tsx [hunk 9]
```diff
@@ -65,13 +124,96 @@ export default function Discover() {
...
+  }, [selectedKeywords, selectedAuthors, selectedSort])
+
+  useEffect(() => {
+    onChange()
+  }, [
+    selectedKeywords.length,
+    selectedAuthors.length,
+    selectedSort.id,
+    onChange,
+  ])
```
Comment (tasola) [inline src/pages/teaching-material/index.tsx:206]: Yeah, it shouldn't be possible to change e.g a keyword without first removing it (changing length) and then adding another one (changing length again), right?

File: src/pages/teaching-material/index.tsx [hunk 10]
```diff
@@ -92,14 +236,29 @@ export default function Discover() {
         </Styled.FilterGroup>
       </div>
       <Styled.H2>All Teaching Material</Styled.H2>
+      <VisuallyHidden id={summaryId} aria-live='polite'>
+        {getResultsSummary()}
+      </VisuallyHidden>
```
Comment (tasola) [inline src/pages/teaching-material/index.tsx:241]: Nice

File: src/types/filters.ts [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    id: 'LEVEL_ASC',
+    label: 'Beginner to Expert',
+    ariaLabel: 'Level from expert to beginner',
+  },
+  levelDESC: {
+    id: 'LEVEL_DESC',
+    label: 'Expert to Beginner',
+    ariaLabel: 'Level from expert to beginner',
+  },
+}
```
Comment (tasola) [inline src/types/filters.ts:55]: Nice file

File: src/components/Chip/Chip.tsx [hunk 12]
```diff
@@ -14,7 +14,7 @@ const Chip = ({ label, id, onDelete }: Props) => {
       {onDelete !== undefined ? (
         <Styled.RemoveButton
           onClick={() => onDelete(id)}
-          aria-label={`Remove chip ${label}`}
+          aria-label={`Remove filter '${label}'`}
```
[the remark this answers is not in the document]
Comment (hakjeri) [reply src/components/Chip/Chip.tsx]: Yes you're right, good catch! I added an optional `type` property which defaults to "chip".

File: src/components/Dropdown/Dropdown.tsx (hunk 3 above)
Comment (hakjeri) [reply src/components/Dropdown/Dropdown.tsx:165]: Well they need different ID:s so a screen reader or other assistive technology doesn't get confused. This part is just for adding supplemental information which will basically add "You have an autocomplete with options available below the input field".

File: src/components/Dropdown/Dropdown.tsx (hunk 4 above)
Comment (hakjeri) [reply src/components/Dropdown/Dropdown.tsx:138]: Yes, I think it's common practice for an input field to just open the options, and not also go down the list.

File: src/pages/teaching-material/index.tsx (hunk 9 above)
Comment (hakjeri) [reply src/pages/teaching-material/index.tsx:203]: Yes, I think just these three will be enough, nothing else really changes the results or their order right now.

**Review by tasola — APPROVED** [review]

**Pull request merged**